staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go

```diff
@@ -0,0 +1,274 @@
+	name := "cr-instance-1"
+	t.Log("Creating a v1 custom resource")
+	crInstance := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+	_, err := customResourceClients["v1"].Create(crInstance, metav1.CreateOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	t.Log("v2 custom resource should be converted")
+	v2cr, err := customResourceClients["v2"].Get(name, metav1.GetOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	if !assert.NotNil(t, v2cr) {
+		t.FailNow()
+	}
+	verifyV2Object(t, crd, v2cr)
+}
+
+func testCRListConversion(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, apiextensionsClient apiextensionsclientset.Interface, customResourceClients map[string]dynamic.ResourceInterface) {
+	name1 := "cr-instance-1"
+	name2 := "cr-instance-2"
+	t.Log("Creating a v1 custom resource")
+
+	// cr-instance-2 is stored as v1. Now change storage version to v2.
+	var updateCRD *apiextensionsv1beta1.CustomResourceDefinition
+	var err error
+	for i := 0; i < 5; i++ {
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get CustomResourceDefinition %q: %v", crd.Name, err)
+		}
+
+		// swap storage version
+		updateCRD.Spec.Versions[0].Storage = false
+		updateCRD.Spec.Versions[1].Storage = true
+
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(updateCRD)
+		if err == nil {
+			break
+		}
+		if !apierrors.IsConflict(err) {
+			break
+		}
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+	crd = updateCRD
+
+	t.Log("Create a v2 custom resource")
+	v1cr3 := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name2,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+
+	// After changing a CRD, the resources for versions will be re-created that can be result in
+	// cancelled connection (e.g. "grpc connection closed" or "context canceled").
+	// Just retrying fixes that.
+	for i := 0; i < 5; i++ {
+		_, err = customResourceClients["v1"].Create(v1cr3, metav1.CreateOptions{})
+		if err == nil {
+			break
+		}
+	}
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	// Now that we have a v1 and v2 object, both list operation in v1 and v2 should work as expected.
+
+	t.Log("List CRs in v1")
+	list, err := customResourceClients["v1"].List(metav1.ListOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	assert.Equal(t, 2, len(list.Items))
+	assert.True(t,
+		(list.Items[0].GetName() == name1 && list.Items[1].GetName() == name2) ||
```

Comment:
testCRListConversion is depending on "cr-instance-1" already having been created in v1?

Reply:
yes